org.h2.mvstore.MVStoreException: Chunk not found [2.2.224/9]

Hi.

While executing a query and "streaming" the content of the `ResultSet` I get a "Chunk XYZ not found" exception.
```
org.h2.jdbc.JdbcSQLNonTransientException: General error: "org.h2.mvstore.MVStoreException: Chunk 19794087 not found [2.2.224/9]" [50000-224]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:566)
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
        at org.h2.message.DbException.get(DbException.java:212)
        at org.h2.message.DbException.convert(DbException.java:407)
        at org.h2.message.DbException.toSQLException(DbException.java:379)
        at org.h2.message.TraceObject.logAndConvert(TraceObject.java:365)
        at org.h2.jdbc.JdbcResultSet.next(JdbcResultSet.java:127)
        at com.zaxxer.hikari.pool.HikariProxyResultSet.next(HikariProxyResultSet.java)
        at com.sixgroup.cit.fmd.eodmaster.cmp.repository.internal.ResultSetStream$1.tryAdvance(ResultSetStream.kt:21)
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
```

Can you please provide some help on why this is happening and how to solve it?

Kind regards and thanks.

I've spoken too early.
Even with a single query and fetchSize = 1 I get the same error from time to time.
> Caused by: org.h2.mvstore.MVStoreException: Chunk 1345210 not found [2.2.224/9]
        at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:996)
        at org.h2.mvstore.FileStore.getChunk(FileStore.java:2006)
        at org.h2.mvstore.FileStore.readPage(FileStore.java:1953)
        at org.h2.mvstore.MVStore.readPage(MVStore.java:1021)
        at org.h2.mvstore.MVMap.readPage(MVMap.java:632)
        at org.h2.mvstore.Page$NonLeaf.getChildPage(Page.java:1117)
        at org.h2.mvstore.Cursor.hasNext(Cursor.java:64)
        at org.h2.mvstore.tx.TransactionMap$CommittedIterator.fetchNext(TransactionMap.java:960)
        at org.h2.mvstore.db.MVSecondaryIndex$MVStoreCursor.next(MVSecondaryIndex.java:434)
        at org.h2.index.IndexCursor.next(IndexCursor.java:292)
        at org.h2.table.TableFilter.next(TableFilter.java:425)
        at org.h2.command.query.Select$LazyResultQueryFlat.fetchNextRow(Select.java:1843)
        at org.h2.result.LazyResult.hasNext(LazyResult.java:78)
        at org.h2.result.FetchedResult.next(FetchedResult.java:34)
        at org.h2.jdbc.JdbcResultSet.nextLazyRow(JdbcResultSet.java:3608)
        at org.h2.jdbc.JdbcResultSet.nextRow(JdbcResultSet.java:3593)
        at org.h2.jdbc.JdbcResultSet.next(JdbcResultSet.java:125)
        ... 43 common frames omitted

The ResultSet is consumed as a Stream with
```kotlin
import org.h2.util.IOUtils.closeSilently
import org.slf4j.LoggerFactory
import java.sql.ResultSet
import java.sql.SQLException
import java.util.Spliterators.AbstractSpliterator
import java.util.function.Consumer
import java.util.stream.Stream
import java.util.stream.StreamSupport.stream
import kotlin.Long.Companion.MAX_VALUE

/**
 * @author Cosimo Damiano Prete
 * @since 09/02/2024
 **/
class ResultSetStream<E>(
    private val resultSet: ResultSet,
    private inline val isClosed: () -> Boolean,
    private inline val rowMapper: (ResultSet) -> E
): Stream<E> by stream(object: AbstractSpliterator<E>(MAX_VALUE, ORDERED or DISTINCT or SORTED or NONNULL or IMMUTABLE) {
    override fun tryAdvance(action: Consumer<in E>) = try {
        val hasNext = !isClosed() && resultSet.next()
        if(hasNext) action.accept(rowMapper(resultSet))

        hasNext
    } catch (ex: SQLException) {
        logger.error(ex.toString(), ex)
        false
    }

    override fun getComparator() = null
}, false) {
    init {
        resultSet.fetchSize = 1
        onClose {
            val statement = resultSet.statement
            val connection = statement.connection

            closeSilently(resultSet)
            closeSilently(statement)
            closeSilently(connection)
        }
    }

    private companion object {
        @JvmStatic
        private val logger = LoggerFactory.getLogger(ResultSetStream::class.java)
    }
}
```

We're seeing the same issue as well, it seems to be exacerbated by running on slow storage mediums such as spinning disks or VMs with virtual storage like Amazon Workspaces when working with large data sets (10M+ rows across several tables, 5+GB database size).  In our case we use H2 as a temporary database that is recreated with a new temp file name every time the application is started, so it shouldn't be a database corruption issue.

One of the instances we've seen recur is the stack trace you have above from running a basic query, though that usually triggers after loading those 10M+ rows, then updating them all a couple of times then querying immediately after.  The updates are done in parallel streams and the query that triggers the error happens immediately after once all updates are complete.

Another variation of this error occurs for us when doing a large `INSERT INTO SELECT FROM` statement that inserts millions of rows into an empty table with no other processing occurring in parallel, no tables involved are being modified by anything else while that's processing:

```
Caused by: org.h2.mvstore.MVStoreException: Chunk 3049 not found [2.2.224/9]
	at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:996)
	at org.h2.mvstore.FileStore.getChunk(FileStore.java:2006)
	at org.h2.mvstore.FileStore.readPage(FileStore.java:1953)
	at org.h2.mvstore.MVStore.readPage(MVStore.java:1021)
	at org.h2.mvstore.MVMap.readPage(MVMap.java:632)
	at org.h2.mvstore.Page$NonLeaf.getChildPage(Page.java:1117)
	at org.h2.mvstore.Cursor.hasNext(Cursor.java:64)
	at org.h2.mvstore.tx.TransactionMap$CommittedIterator.fetchNext(TransactionMap.java:960)
	at org.h2.mvstore.db.MVSecondaryIndex$MVStoreCursor.next(MVSecondaryIndex.java:434)
	at org.h2.index.IndexCursor.next(IndexCursor.java:292)
	at org.h2.table.TableFilter.next(TableFilter.java:425)
	at org.h2.command.dml.FilteredDataChangeStatement.nextRow(FilteredDataChangeStatement.java:71)
	at org.h2.command.dml.Delete.update(Delete.java:58)
	at org.h2.command.dml.DataChangeStatement.update(DataChangeStatement.java:74)
	at org.h2.command.CommandList.executeRemaining(CommandList.java:49)
	at org.h2.command.CommandList.update(CommandList.java:66)
	at org.h2.command.Command.executeUpdate(Command.java:256)
```

Our code base is quite large and I've tried to create a simplified example to repro the issue but so far haven't had much success.  Reproducing it in our code base is still intermittent and takes hours of processing to trigger.  The most success I've had is plugging in an old spinning disk and moving the DB to that drive for testing.

> We're seeing the same issue as well, it seems to be exacerbated by running on slow storage mediums such as spinning disks or VMs with virtual storage like Amazon Workspaces when working with large data sets (10M+ rows across several tables, 5+GB database size). In our case we use H2 as a temporary database that is recreated with a new temp file name every time the application is started, so it shouldn't be a database corruption issue.
> 
> One of the instances we've seen recur is the stack trace you have above from running a basic query, though that usually triggers after loading those 10M+ rows, then updating them all a couple of times then querying immediately after. The updates are done in parallel streams and the query that triggers the error happens immediately after once all updates are complete.
> 
> Another variation of this error occurs for us when doing a large `INSERT INTO SELECT FROM` statement that inserts millions of rows into an empty table with no other processing occurring in parallel, no tables involved are being modified by anything else while that's processing:
> 
> ```
> Caused by: org.h2.mvstore.MVStoreException: Chunk 3049 not found [2.2.224/9]
> 	at org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:996)
> 	at org.h2.mvstore.FileStore.getChunk(FileStore.java:2006)
> 	at org.h2.mvstore.FileStore.readPage(FileStore.java:1953)
> 	at org.h2.mvstore.MVStore.readPage(MVStore.java:1021)
> 	at org.h2.mvstore.MVMap.readPage(MVMap.java:632)
> 	at org.h2.mvstore.Page$NonLeaf.getChildPage(Page.java:1117)
> 	at org.h2.mvstore.Cursor.hasNext(Cursor.java:64)
> 	at org.h2.mvstore.tx.TransactionMap$CommittedIterator.fetchNext(TransactionMap.java:960)
> 	at org.h2.mvstore.db.MVSecondaryIndex$MVStoreCursor.next(MVSecondaryIndex.java:434)
> 	at org.h2.index.IndexCursor.next(IndexCursor.java:292)
> 	at org.h2.table.TableFilter.next(TableFilter.java:425)
> 	at org.h2.command.dml.FilteredDataChangeStatement.nextRow(FilteredDataChangeStatement.java:71)
> 	at org.h2.command.dml.Delete.update(Delete.java:58)
> 	at org.h2.command.dml.DataChangeStatement.update(DataChangeStatement.java:74)
> 	at org.h2.command.CommandList.executeRemaining(CommandList.java:49)
> 	at org.h2.command.CommandList.update(CommandList.java:66)
> 	at org.h2.command.Command.executeUpdate(Command.java:256)
> ```
> 
> Our code base is quite large and I've tried to create a simplified example to repro the issue but so far haven't had much success. Reproducing it in our code base is still intermittent and takes hours of processing to trigger. The most success I've had is plugging in an old spinning disk and moving the DB to that drive for testing.

Too bad that on my side I've the DB file on an SSD :-/

--- a/h2/src/main/org/h2/command/CommandList.java
+++ b/h2/src/main/org/h2/command/CommandList.java
@@ -20,7 +20,7 @@ import org.h2.result.ResultWithGeneratedKeys;
  */
 class CommandList extends Command {
 
-    private CommandContainer command;
+    private final CommandContainer command;
     private final ArrayList<Prepared> commands;
     private final ArrayList<Parameter> parameters;
     private String remaining;
@@ -42,20 +42,21 @@ class CommandList extends Command {
 
     private void executeRemaining() {
         for (Prepared prepared : commands) {
+            CommandContainer commandContainer = new CommandContainer(session, prepared.getSQL(), prepared);
             prepared.prepare();
             if (prepared.isQuery()) {
-                prepared.query(0);
+                executeQuery(0, false);
             } else {
-                prepared.update();
+                commandContainer.executeUpdate(null);
             }
         }
         if (remaining != null) {
             remainingCommand = session.prepareLocal(remaining);
             remaining = null;
             if (remainingCommand.isQuery()) {
-                remainingCommand.query(0);
+                remainingCommand.executeQuery(0, false);
             } else {
-                remainingCommand.update(null);
+                remainingCommand.executeUpdate(null);
             }
         }
     }

